### DISCUSSION TITLE
Model card design and content

### DISCUSSION THREAD

**Discussion by ddbuscombe** [Doodleverse/segmentation_zoo#3] [General]
(this could/should be a Discussion, but that doesn't appear to be an option??)

I favour a simple text description of the models such as Fig 1 in https://arxiv.org/pdf/1810.03993.pdf in markdown format

@ebgoldstein could you please help add / edit the info fields below for a Zoo model (I will revisit Fri)

## Model Details.
> Basic info
* Person or organization developing model
* Model date
* Model version
* Model type
> Model files and their meaning / uses
* json file containing model architecture
* h5 file containing trained weights
* model card (markdown format)
* 
> Information about training algorithms, parameters, fair-
ness constraints or other applied approaches, and features

> Paper or other resource for more information
* Citation details
* License
* Where to send questions or comments about the model - could be this Issues page

## Intended Use. 
Use cases that were envisioned during de-
velopment.
* Primary intended uses: image segmentation  with the intended image source and class list
* Primary intended users: Zoo users / transfer learning 
* Out-of-scope use cases:  whole image classification (does an image contain a class?) and object detection (masks can be converted into bounding boxes around each class)

## Factors. Factors could include demographic or phenotypic
groups, environmental conditions, technical attributes, or
others listed in Section 4.3.
> Relevant factors
* 
> Evaluation factors
* IoU, Dice, KLD

## Metrics. Metrics should be chosen to reflect potential real-
world impacts of the model.
> Model performance measures
* ?
> Decision thresholds
* ?
> Variation approaches
* ?

## Evaluation Data. Details on the dataset(s) used for the
quantitative analyses in the card.
> Datasets
* ?
> Motivation
* ?
> Preprocessing
* ?

## Training Data
> description of training data
 * how many labeled images
 * how many classes / what classes
 * source of imagery
 * source of labels
> link to dataset
* e.g. C/T

• Quantitative Analyses
> Unitary results
* ??
> Intersectional results 
* ??

## Ethical Considerations
* ??

## Caveats and Recommendations
* A brief description of how to use this model with Segmentation Gym functions?

**Comment by ebgoldstein** [comment]
i will give it a think.. and discussions is working now...

**Reply by ddbuscombe** [reply]
thanks, we are now a Discussion. 

from [hugginface](https://huggingface.co/docs/hub/model-repos), [this format](https://raw.githubusercontent.com/huggingface/huggingface_hub/main/modelcard.md) seems useful and is similar to the usgs metadata reporting

**Comment by ddbuscombe** [comment]
The config file is essentially the model card. The weights must be shipped with the json file, and the program needs both files for inference. However, it does not contain certain fields such as the name of the dataset it is based on. 

A modified json version of the config file seems to work ok for our purposes?

```
{
	"DETAILS": {
		"NAME": "Segmentation Zoo models for Landsat-8 satellite imagery, Coast Train v1 Landsat-8 4-class subset.",
    "DATE": "2022-02-22",
    "URL":"10.5281/zenodo.6227988",
    "CITATION":"Buscombe, D. (2022) Segmentation Zoo models for Landsat-8 satellite imagery, Coast Train v1 Landsat-8 4-class subset. Zenodo data release 10.5281/zenodo.6227988",
    "QUERIES": "https://github.com/Doodleverse/segmentation_zoo/issues",
    "CREDIT":"Daniel Buscombe, @dbuscombe-usgs",
    "INTENDED_PURPOSE":"Landsat-8 satellite image segmentation",
    "KEYWORDS": {
      "1": "Landsat",
      "2": "Landsat-8",
			"3": "water",
			"4": "whitewater",
			"5": "sediment",
			"6": "other"
    }
	},  
	"DATASET": {
		"NAME": "Coast Train v1 Landsat-8", 
    "SOURCE": "https://cmgds.marine.usgs.gov/data-releases/datarelease/10.5066-P91NP87I",
    "CITATION": "Wernette, P.A., Buscombe, D.D., Favela, J., Fitzpatrick, S., and Goldstein E., 2022, Coast Train--Labeled imagery for training and evaluation of data-driven models for image segmentation: U.S. Geological Survey data release, https://doi.org/10.5066/P91NP87I",
    "NUMBER_LABELED_IMAGES": 419,
		"CLASSES": {
			"0": "null",
			"1": "water",
			"2": "whitewater",
			"3": "sediment",
			"4": "other"
		},
		"ORIG_CLASSES": {
			"0": "water",
			"1": "whitewater",
			"2": "sediment",
			"3": "other_natural_terrain",
			"4": "vegetated_surface",
			"5": "agricultural",
			"6": "development",
			"7": "cloud",
			"8": "nodata",
			"9": "unusual",
			"10": "unknown"
		},
		"REMAP_CLASSES": {
			"0": 0,
			"1": 1,
			"2": 2,
			"3": 3,
			"4": 4,
			"5": 4,
			"6": 4,
			"7": 4,
			"8": 4,
			"9": 4,
			"10": 4,
			"11": 4
		},
		"N_DATA_BANDS": 1
	},
	"MODEL": {
		"NAME": "resunet",
    "KERNEL":7,
    "STRIDE":2,
    "FILTERS":6    
	},
	"TRAINING": {
    "BATCH_SIZE": 8,
    "DROPOUT":0.1,
    "DROPOUT_CHANGE_PER_LAYER":0.0,
    "DROPOUT_TYPE":"standard",
    "USE_DROPOUT_ON_UPSAMPLING":false,
    "LOSS":"cat",
    "PATIENCE": 10,
    "MAX_EPOCHS": 100,
    "VALIDATION_SPLIT": 0.6,
    "RAMPUP_EPOCHS": 20,
    "SUSTAIN_EPOCHS": 0.0,
    "EXP_DECAY": 0.9,
    "START_LR":  1e-7,
    "MIN_LR": 1e-7,
    "MAX_LR": 1e-4   
	},
	"AUGMENTATION": {
    "AUGMENTATION_USED": true,
    "FILTER_VALUE": 0,
    "AUG_ROT": 0.05,
    "AUG_ZOOM": 0.05,
    "AUG_WIDTHSHIFT": 0.05,
    "AUG_HEIGHTSHIFT": 0.05,
    "AUG_HFLIP": true,
    "AUG_VFLIP": false,
    "AUG_LOOPS": 10,
    "AUG_COPIES": 5  
	}     
}

```

**Comment by ddbuscombe** [comment]
Generic example

```
{
	"DETAILS": {
	"NAME": "name of model set",
    "DATE": "YYYY-MM-DD",
    "URL":"string",
    "CITATION":"Bloggs, J. (2022) Name. Zenodo data release XXXXXXX",
    "QUERIES": "email or website",
    "CREDIT":"Name",
    "INTENDED_PURPOSE":"Description",
    "KEYWORDS": {
      "1": "keyword 1",
      "2": "keyword 2"
    }
	},  
	"DATASET": {
	"NAME": "Dataset name", 
    "SOURCE": "dataset url or description of how to obtain data",
    "CITATION": "Bloggs et al 2022, Name of dataset: Publisher, URL",
    "NUMBER_LABELED_IMAGES": 999999,
		"CLASSES": {
			"0": "class 1",
			"1": "class 2",
			"2": "class 3"
		},
		"ORIG_CLASSES": {
			"0": "orig class 1",
			"1": "orig class 2",
			"2": "orig class 3",
			"3": "orig class 4"
		},
		"REMAP_CLASSES": {
			"0": 0,
			"1": 2,
			"2": 4,
			"3": 6
		},
		"N_DATA_BANDS": 3,
    "BAND_NAMES": {
      "0": "red",
      "1": "green",
      "2": "blue"
    }
	},
	"MODEL": {
	"NAME": "resunet",
    "KERNEL":7,
    "STRIDE":2,
    "FILTERS":6    
	},
	"TRAINING": {
    "BATCH_SIZE": 8,
    "DROPOUT":0.1,
    "DROPOUT_CHANGE_PER_LAYER":0.0,
    "DROPOUT_TYPE":"standard",
    "USE_DROPOUT_ON_UPSAMPLING":false,
    "LOSS":"cat",
    "PATIENCE": 10,
    "MAX_EPOCHS": 100,
    "VALIDATION_SPLIT": 0.6,
    "RAMPUP_EPOCHS": 20,
    "SUSTAIN_EPOCHS": 0.0,
    "EXP_DECAY": 0.9,
    "START_LR":  1e-7,
    "MIN_LR": 1e-7,
    "MAX_LR": 1e-4   
	},
	"AUGMENTATION": {
    "AUGMENTATION_USED": true,
    "FILTER_VALUE": 0,
    "AUG_ROT": 0.05,
    "AUG_ZOOM": 0.05,
    "AUG_WIDTHSHIFT": 0.05,
    "AUG_HEIGHTSHIFT": 0.05,
    "AUG_HFLIP": true,
    "AUG_VFLIP": false,
    "AUG_LOOPS": 10,
    "AUG_COPIES": 5  
	}     
}

```

**Comment by ddbuscombe** [comment]
I included the first version of this idea in the first Coast-Train related model release https://zenodo.org/record/6229071#.YhVhafuIZhE. Keeping for now but can already think of things to add, such as details about target imagery (bands, formats, dimensions, etc)

**Comment by ddbuscombe** [comment]
A simple utility to create a modelcard would be handy. The input would be the config file that made the model, and either another one that describes the dataset used, or a series of prompts

Where possible, the images and labels used to make the model should also be included in the zenodo releases
